File: CRC.java
//CRC Checksum
import java.util.*;

public class CRC{
    public static void main(String[] args){
        Scanner in = new Scanner(System.in);
        System.out.print("Enter message bits: ");
        String message = in.nextLine();
        System.out.print("Enter the generator: ");
        String generator = in.nextLine();

        int[] data = new int[(2*message.length())-generator.length()+1];
        int[] divisor = new int[generator.length()];
        int[] checksum = new int[(2*message.length())-generator.length()+1];

        for(int i=0;i<message.length();i++)
            data[i] = checksum[i] = Integer.parseInt(message.charAt(i)+"");
        for(int i=message.length();i<data.length;i++)
            data[i]=0;
        for(int i=0;i<generator.length();i++)
                        divisor[i] = Integer.parseInt(generator.charAt(i)+"");

        //Calculation of CRC
        for(int i=0;i<=(data.length-divisor.length);i++){
            if(data[i]==1)
                for(int j=0;j<divisor.length;j++)
                    data[i+j]^=divisor[j];
        }
        for(int i=message.length();i<data.length;i++)
                        checksum[i] = data[i];
        //Display of CRC
        System.out.print("The Checksum code is: ");
        for(int i=0;i<data.length;i++)
            System.out.print(checksum[i]);
        System.out.println();
        //Check for input CRC code
            System.out.print("Enter the recieved Checksum code: ");
                message = in.nextLine();
                
                data = new int[message.length()];
                
                for(int i=0;i<message.length();i++)
                        data[i] = Integer.parseInt(message.charAt(i)+"");
        //Calculation of remainder
        for(int i=0;i<=(data.length-divisor.length);i++){
                        if(data[i]==1)
                                for(int j=0;j<divisor.length;j++)
                                        data[i+j]^=divisor[j];
                }
        //Check for Validity
        boolean valid = true;
        for(int i=0;i<data.length;i++){
                        if(data[i]==1){
                valid = false;
                break;
            }
        }
        if(valid==true)
            System.out.println("Data stream is valid");
        else
            System.out.println("Data stream is invalid. CRC error occurred.");
        in.close();
    }
}

File: crc.cpp
#include<bits//stdc++.h>
using namespace std;

string crc(string data, string poly, bool errChk){
    string rem = data;
    if(!errChk){
        for(int i = 0; i < poly.length()-1; i++)
            rem.append("0");
    }
    for(int i = 0; i < rem.length()-poly.length()+1; i++){
        if(rem[i] == '1'){
            for(int j = 0; j < poly.length(); j++){
                if(rem[i+j] == poly[j]){
                    rem[i+j] = '0';
                }
                else{
                    rem[i+j] = '1';
                }
            }
        }
    }
    return rem.substr(rem.length()-poly.length()+1);
}   

int main(){
    string data, poly;
    cout << "Enter data to be sent: ";
    cin >> data;

    cout << "Enter gererating polynomial : ";
    cin >> poly;

    string rem = crc(data, poly, 0);
    string codeword = data+rem;
    cout << "Remainder : " << rem << endl;
    cout << "Codeword : " << codeword << endl;

    //Error checking
    string newCodeword;
    cout << "Enter data that is recieved : ";
    cin >> newCodeword;
    string newRem = crc(newCodeword, poly, 1);
    if(stoi(newRem) == 0)
        cout << "No error in data transmission" << endl;
    else    
        cout << "Error in data transmission" << endl;

}
